src/libraries/Common/src/Interop/OSX/Swift.Runtime/CryptoKit.cs

```diff
@@ -0,0 +1,274 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Runtime.InteropServices.Swift;
+
+#pragma warning disable CS3016 // Arrays as attribute arguments are not CLS Compliant
+#pragma warning disable SYSLIB1051
+
+namespace Swift.Runtime
+{
+    /// <summary>
+    /// Represents ChaChaPoly in C#.
+    /// </summary>
+    internal unsafe partial struct ChaChaPoly
+    {
+        /// <summary>
+        /// Represents Nonce in C#.
+        /// </summary>
+        internal sealed unsafe partial class Nonce
+        {
+            private const int _payloadSize = 16;
+            internal void* payload;
+
+            internal Nonce()
+            {
+                payload = Marshal.AllocHGlobal(_payloadSize).ToPointer();
+                SwiftIndirectResult swiftIndirectResult = new SwiftIndirectResult(payload);
+                CryptoKit.PInvoke_ChaChaPoly_Nonce_Init(swiftIndirectResult);
+            }
+
+            internal Nonce(Data data)
+            {
+                payload = Marshal.AllocHGlobal(_payloadSize).ToPointer();
+                SwiftIndirectResult swiftIndirectResult = new SwiftIndirectResult(payload);
+                void* dataPtr = &data;
+
+                CryptoKit.PInvoke_ChaChaPoly_Nonce_Init2(swiftIndirectResult, dataPtr, data.Metadata(), default(DataProtocol).WitnessTable(data), out SwiftError error);
+
+                if (error.Value != null) // Should throw when nonce is not 12 bytes
```

Comment:
`Value` [is a pointer](https://github.com/dotnet/runtime/blob/86995059adb5e76e1701502b980c0bf574451a94/src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Swift/SwiftTypes.cs#L108). I recall that roslyn did not do the right thing for pattern matching and pointers at some point, but perhaps it's been fixed since.